tensorly/decomposition/tests/test_parafac2.py

```diff
@@ -0,0 +1,172 @@
+        shapes=[(50, 60) for _ in range(40)],
+        rank=rank,
+        random_state=1234
+    )
+    tensor = parafac2_to_tensor(random_parafac2_tensor)
+    slices = parafac2_to_slices(random_parafac2_tensor)
+
+    slice_rec = parafac2(slices, rank, random_state=1234, normalize_factors=False)
+    slice_rec_tensor = parafac2_to_tensor(slice_rec)
+
+    tensor_rec = parafac2(tensor, rank, random_state=1234, normalize_factors=False)
+    tensor_rec_tensor = parafac2_to_tensor(tensor_rec)
+
+    assert tl.max(tl.abs(slice_rec_tensor - tensor_rec_tensor)) < 1e-8
+
+
+def test_parafac2_normalize_factors():
+    rng = check_random_state(1234)
+    rank = 6
+
+    random_parafac2_tensor = random_parafac2(
+        shapes=[(50 + rng.randint(10),60) for _ in range(40)],
+        rank=rank,
+        random_state=rng,
+    )
+    random_parafac2_tensor.weights[0] = 100
+    slices = parafac2_to_tensor(random_parafac2_tensor)
+    normalized_rec = parafac2(slices, rank, random_state=rng, normalize_factors=True)
+    assert normalized_rec.weights[0] > 1
+    assert tl.max(tl.abs(T.norm(normalized_rec.factors[0], axis=0) - 1)) < 1e-5
+    
+    unnormalized_rec = parafac2(slices, rank, random_state=rng, normalize_factors=False)
+    assert unnormalized_rec.weights[0] == 1
+
+
+def test_parafac2_init_valid():
+    rng = check_random_state(1234)
+    rank = 6
+
+    random_parafac2_tensor = random_parafac2(shapes=[(30,60)]*50, rank=rank, random_state=rng)
+    tensor = parafac2_to_tensor(random_parafac2_tensor)
+    weights, (A, B, C), projections = random_parafac2_tensor
+    B = T.dot(projections[0], B)
+    
+    for init_method in ['random', 'svd', random_parafac2_tensor, (weights, (A, B, C))]:
+        init = initialize_decomposition(tensor, rank, init=init_method)
+        assert init.shape == random_parafac2_tensor.shape
+
+
+def test_parafac2_init_error():
+    rng = check_random_state(1234)
+    rank = 6
+
+    random_parafac2_tensor = random_parafac2(shapes=[(30,60)]*50, rank=rank, random_state=rng)
+    tensor = parafac2_to_tensor(random_parafac2_tensor)
+
+    with np.testing.assert_raises(ValueError):
+        _ = initialize_decomposition(tensor, rank, init='bogus init type')
+
+    with np.testing.assert_raises(ValueError):
+        _ = initialize_decomposition(tensor, rank, init=('another', 'bogus', 'init', 'type'))
+
+def test_parafac2_to_tensor():
+    rng = check_random_state(1234)
+    rank = 6
+
+    I = 50
+    J = 30
+    K = 60
+
+    weights, factors, projections = random_parafac2(shapes=[(J, K)]*I, rank=rank, random_state=rng)
+
+    constructed_tensor = parafac2_to_tensor((weights, factors, projections))
+    tensor_manual = T.zeros((I, J, K), **T.context(weights))
+
+    for i in range(I):
+        Bi = T.dot(projections[i], factors[1])
+        for j in range(J):
+            for k in range(K):
+                for r in range(rank):
+                    tensor_manual[i, j, k] += factors[0][i][r]*Bi[j][r]*factors[2][k][r]
```

Comment:
We can also just use index_update here, should be something like:
```suggestion
                    tl.index_update(tensor_manual, index[i, j, k],  tensor_manual[i, j, k] + factors[0][i][r]*Bi[j][r]*factors[2])[k][r]
```